staticjinja/dep_graph.py

```diff
@@ -0,0 +1,80 @@
+# -*- coding:utf-8 -*-
+
+"""
+Dependency graph for staticjina
+"""
+
+from copy import deepcopy
+
+
+class DepGraph(object):
+    """
+    A directed graph which will handle dependencies between templates and data
+    files in a site.
+
+    :param site:
+        A :class:`Site <Site>` object.
+
+    """
+    def __init__(self, site):
+        self.parents = dict(
+                (s, set()) for s in site.jinja_names + site.data_names
+                )
+        self.children = deepcopy(self.parents)
+
+        for filename in site.jinja_names:
+            self.parents[filename] = site.get_file_dep(filename)
+            for d in self.parents[filename]:
+                self.children[d].add(filename)
+
+        self.site = site
+
+    def connected_components(self, adjacency, start):
+        """Returns the (directed) connected component of start in the graph with
+        given adjacency dict.
+
+        Uses a classical depth first search.
+
+        :param adjacency: the adjacency dict to be used (will be either
+        self.children or self.parents in all foreseen uses)
+
+        :param start: the vertex in the graph whose connected component we
+        seek.
+        """
+
+        seen = set([start])
+        stack = [iter(adjacency[start])]
+        while stack:
+            children = stack[-1]
+            try:
+                child = next(children)
+                if child not in seen:
+                    yield child
+                seen.add(child)
+                stack.append(iter(adjacency[child]))
```

Comment:
This algorithm will infinitely recurse if there is a cycle in the graph because of this line. (Consider the case where the graph is a single node connected to itself.) Did you mean to indent these two lines, or is this algorithm supposed to only work on trees?